{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fa4d98b9-a59d-448d-ba31-fd2021b32298",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 设计TE的探针，不再进行过滤，而是进行打分\n",
    "## 只用全长进行比对，即62bp；\n",
    "## 首先和基因组进行比对，使用BLASTN，然后和rmsk进行比对，使用Bowtie2。\n",
    "## 接受多条输入"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d9cb1d9c-6bf8-44f0-9e51-4a7bbefd80ed",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "TE_INPUT = {\n",
    "   'Lx': [\n",
    "        'chr1|129376168|129376548|+'\n",
    "   ],\n",
    "    \n",
    "    'IAPLTR3-int': [\n",
    "        'chr2|85788757|85789706|+'\n",
    "   ],\n",
    "    \n",
    "    'Lx2B': [\n",
    "        'chr6|21242260|21243219|-',\n",
    "        'chr8|33101938|33102145|+'\n",
    "   ],\n",
    "    \n",
    "    'L1Lx_II': [\n",
    "        'chr1|32424978|32425767|-',\n",
    "   ],\n",
    "    \n",
    "    'RLTR4_Mm': [\n",
    "        'chr4|133886950|133887248|+',\n",
    "        'chr4|133891467|133891737|+'\n",
    "   ],\n",
    "    \n",
    "    'RLTR44-int': [\n",
    "        'chr11|34541427|34541644|+',\n",
    "        'chr11|34551131|34551356|+'\n",
    "   ],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "afa2c7b9-3a15-4f82-8a64-a574b24447e6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(TE_INPUT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5cdf701d-91d7-44f3-a6d9-603dc209b72d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['Lx', 'IAPLTR3-int', 'Lx2B', 'L1Lx_II', 'RLTR4_Mm', 'RLTR44-int'])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TE_INPUT.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2e7fd9c-78e0-4844-a4c9-996639cdde42",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 1.生成candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "fe7feae1-b3e4-4504-95e4-6d75a9f112f0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "key = 'IAPLTR3-int'\n",
    "bridge_id = 'Pad_1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df4191da-141c-4985-aa8d-a97a591a2fe8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "5ed04dc1-90c7-4e3f-b57a-d90aadcaa738",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pyfaidx import Fasta\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "def get_genome_seq(chr_name:str, start:int, end:int, strand: str) -> SeqRecord:\n",
    "    \n",
    "    mouse_fasta_path = \"/media/duan/sda2/Reference_DataBase/mm39/GRCm39.genome.fa\"\n",
    "    \n",
    "    fasta_ref = Fasta(mouse_fasta_path)\n",
    "\n",
    "    target = fasta_ref[chr_name][start:end]\n",
    "    \n",
    "    if strand == \"-\":\n",
    "        target = target.complement\n",
    "    \n",
    "    return SeqRecord(seq = Seq(target.seq), id = f\"{chr_name}|{start}|{end}\", description=strand)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "80b7ecf3-c731-4c13-b38a-b611d40ef42f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sequences = []\n",
    "for location in TE_INPUT[key]:\n",
    "    chrom, start, end, strand = location.split(\"|\")\n",
    "    sequences.append(get_genome_seq(chrom, int(start), int(end), strand))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "5c352611-7f7c-4c8b-81ff-b6d59327a9e8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "030dda87-2bd3-4180-8e5f-d1ee059dab57",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqRecord(seq=Seq('CTGGTGCCGAAGATCCCGGGAATTCTACAACATCGCCGGCATGGAGGAGGTCCC...GAG'), id='chr2|85788757|85789706', name='<unknown name>', description='+', dbxrefs=[])]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "0a0bd1d2-83c8-4d07-8c45-6aa3dba2e055",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Seq import reverse_complement\n",
    "from collections import Counter\n",
    "\n",
    "def findAllCandidates(seqs: list):\n",
    "    \n",
    "    All_probes = []\n",
    "    ALL_probes_count = Counter()\n",
    "    \n",
    "    for target in seqs:\n",
    "        limit = len(target)\n",
    "        for i in range(0, limit-62):\n",
    "            \n",
    "            start = i\n",
    "            end = i + 62\n",
    "            target_region = target.seq[start:end].upper()\n",
    "            \n",
    "            if target.description == '-':\n",
    "                \n",
    "                probe = SeqRecord(target_region, id = f\"{target.id}_{start}_plus\", description=str(start))\n",
    "                before_len = len(ALL_probes_count)\n",
    "                ALL_probes_count[str(target_region)] += 1\n",
    "                after_len = len(ALL_probes_count)\n",
    "                if before_len != after_len:\n",
    "                    All_probes.append(probe)\n",
    "            \n",
    "            if target.description == '+':\n",
    "                \n",
    "                target_region = target_region.reverse_complement()\n",
    "                probe = SeqRecord(target_region, id = f\"{target.id}_{start}_minus\", description=str(start))\n",
    "                before_len = len(ALL_probes_count)\n",
    "                ALL_probes_count[str(target_region)] += 1\n",
    "                after_len = len(ALL_probes_count)\n",
    "                if before_len != after_len:\n",
    "                    All_probes.append(probe)\n",
    "                \n",
    " \n",
    "    prbs1 = []\n",
    "    prbs2 = []\n",
    "    \n",
    "    for probe in All_probes:\n",
    "        \n",
    "        p1_rec = probe[32:62]\n",
    "        p2_rec = probe[0:30]\n",
    "        \n",
    "        p1_rec.id = probe.id + '_p1'\n",
    "        p2_rec.id = probe.id + '_p2'\n",
    "        \n",
    "        prbs1.append(p1_rec)\n",
    "        prbs2.append(p2_rec)\n",
    "    \n",
    "    return pd.DataFrame({\"Region\": All_probes, \"Probe1\":prbs1, \"Probe2\":prbs2}), ALL_probes_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "ad16aaba-e545-481c-91eb-eaf38afaf0a8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "Probes, probes_counts = findAllCandidates(seqs = sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "d26b9ae8-916a-45cb-83fa-953acef14000",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "887"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Probes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dafcfce3-e1bd-41ad-92a4-6b1ce89f71eb",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 2.进行Basic过滤"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "db109f2e-e98a-479e-b3d2-714c30e999d9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pandarallel import pandarallel\n",
    "pandarallel.initialize(progress_bar = False, verbose=0, nb_workers = 16)\n",
    "\n",
    "from Bio.SeqUtils import MeltingTemp\n",
    "from Bio.Seq import Seq\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math\n",
    "from collections import Counter\n",
    "import pysam as ps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "e029b333-8c05-4a01-853e-c2344e5da30b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_tm(seq):\n",
    "    \n",
    "    tmp = MeltingTemp.Tm_NN(seq, nn_table=MeltingTemp.R_DNA_NN1, dnac1 = 20, \n",
    "                            selfcomp = True, Na = 390, Mg = 0)\n",
    "    \n",
    "    GC = (seq.count(\"G\") + seq.count(\"C\"))/len(seq) * 100\n",
    "    \n",
    "    return MeltingTemp.chem_correction(tmp, fmd=30 * 0.2222, fmdmethod=2, GC=GC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "a0cede3c-c603-44be-96c5-66181f827397",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def basicFilter(probe):\n",
    "    \n",
    "    probe = str(probe.seq)\n",
    "    \n",
    "    # filter Ns\n",
    "    if \"N\" in probe:\n",
    "        return \"FAILED\"\n",
    "\n",
    "    base_count = Counter(probe)\n",
    "    total_bases = len(probe)\n",
    "    \n",
    "    # Filter Complexity\n",
    "    # Shannon entropy\n",
    "    frequencies = {base: count / total_bases for base, count in base_count.items()}\n",
    "    entropy = -sum(freq * math.log2(freq) for freq in frequencies.values() if freq > 0)\n",
    "    \n",
    "    if entropy < 1:\n",
    "        return \"FAILED\"\n",
    "    \n",
    "    # filter base percentage\n",
    "    GC = (base_count['G'] + base_count['C'])/total_bases * 100\n",
    "    if GC < 20 or GC > 80:\n",
    "        return \"FAILED\"\n",
    "    \n",
    "    # filter repeats\n",
    "    REPEATS_LIST = [\n",
    "        \"GGGGG\", \"CCCCC\"\n",
    "    ]\n",
    "    \n",
    "    repeats_count = 0\n",
    "    for BL in REPEATS_LIST:\n",
    "        if probe.count(BL) > 0:\n",
    "            return \"FAILED\"\n",
    "    \n",
    "    # filter tm\n",
    "    tm = get_tm(probe)\n",
    "    if tm < 42:\n",
    "        return \"FAILED\"\n",
    "    \n",
    "            \n",
    "    return \"PASS\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "70295128-6aa2-46ab-809d-21cb7ef3bbb6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "Probes[\"basicFilter1\"] = Probes[\"Probe1\"].parallel_apply(lambda x:basicFilter(x))\n",
    "Probes[\"basicFilter2\"] = Probes[\"Probe2\"].parallel_apply(lambda x:basicFilter(x))\n",
    "Probes = Probes[(Probes[\"basicFilter1\"] == \"PASS\") & (Probes[\"basicFilter2\"] == \"PASS\")].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "1455ba13-32ae-422e-982b-343f005e7869",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "551"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Probes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "e4face8a-5401-41bf-b291-f912fa02afc8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>Probe1</th>\n",
       "      <th>Probe2</th>\n",
       "      <th>basicFilter1</th>\n",
       "      <th>basicFilter2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>(T, G, C, A, G, T, T, C, T, G, A, A, C, C, C, ...</td>\n",
       "      <td>(C, C, T, C, C, T, C, C, A, T, G, C, C, G, G, ...</td>\n",
       "      <td>(T, G, C, A, G, T, T, C, T, G, A, A, C, C, C, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>(C, T, G, C, A, G, T, T, C, T, G, A, A, C, C, ...</td>\n",
       "      <td>(A, C, C, T, C, C, T, C, C, A, T, G, C, C, G, ...</td>\n",
       "      <td>(C, T, G, C, A, G, T, T, C, T, G, A, A, C, C, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>(A, C, T, G, C, A, G, T, T, C, T, G, A, A, C, ...</td>\n",
       "      <td>(G, A, C, C, T, C, C, T, C, C, A, T, G, C, C, ...</td>\n",
       "      <td>(A, C, T, G, C, A, G, T, T, C, T, G, A, A, C, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>(C, A, C, T, G, C, A, G, T, T, C, T, G, A, A, ...</td>\n",
       "      <td>(G, G, A, C, C, T, C, C, T, C, C, A, T, G, C, ...</td>\n",
       "      <td>(C, A, C, T, G, C, A, G, T, T, C, T, G, A, A, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>(C, C, A, C, T, G, C, A, G, T, T, C, T, G, A, ...</td>\n",
       "      <td>(G, G, G, A, C, C, T, C, C, T, C, C, A, T, G, ...</td>\n",
       "      <td>(C, C, A, C, T, G, C, A, G, T, T, C, T, G, A, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809</th>\n",
       "      <td>(A, T, G, A, G, C, T, C, C, T, T, A, T, C, T, ...</td>\n",
       "      <td>(C, T, G, G, C, A, T, G, C, T, G, T, C, T, T, ...</td>\n",
       "      <td>(A, T, G, A, G, C, T, C, C, T, T, A, T, C, T, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>810</th>\n",
       "      <td>(T, A, T, G, A, G, C, T, C, C, T, T, A, T, C, ...</td>\n",
       "      <td>(T, C, T, G, G, C, A, T, G, C, T, G, T, C, T, ...</td>\n",
       "      <td>(T, A, T, G, A, G, C, T, C, C, T, T, A, T, C, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811</th>\n",
       "      <td>(T, T, A, T, G, A, G, C, T, C, C, T, T, A, T, ...</td>\n",
       "      <td>(T, T, C, T, G, G, C, A, T, G, C, T, G, T, C, ...</td>\n",
       "      <td>(T, T, A, T, G, A, G, C, T, C, C, T, T, A, T, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>(T, T, T, A, T, G, A, G, C, T, C, C, T, T, A, ...</td>\n",
       "      <td>(T, T, T, C, T, G, G, C, A, T, G, C, T, G, T, ...</td>\n",
       "      <td>(T, T, T, A, T, G, A, G, C, T, C, C, T, T, A, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816</th>\n",
       "      <td>(T, C, T, T, T, T, T, A, T, G, A, G, C, T, C, ...</td>\n",
       "      <td>(T, C, A, G, T, T, T, C, T, G, G, C, A, T, G, ...</td>\n",
       "      <td>(T, C, T, T, T, T, T, A, T, G, A, G, C, T, C, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>551 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Region  \\\n",
       "20   (T, G, C, A, G, T, T, C, T, G, A, A, C, C, C, ...   \n",
       "21   (C, T, G, C, A, G, T, T, C, T, G, A, A, C, C, ...   \n",
       "22   (A, C, T, G, C, A, G, T, T, C, T, G, A, A, C, ...   \n",
       "23   (C, A, C, T, G, C, A, G, T, T, C, T, G, A, A, ...   \n",
       "24   (C, C, A, C, T, G, C, A, G, T, T, C, T, G, A, ...   \n",
       "..                                                 ...   \n",
       "809  (A, T, G, A, G, C, T, C, C, T, T, A, T, C, T, ...   \n",
       "810  (T, A, T, G, A, G, C, T, C, C, T, T, A, T, C, ...   \n",
       "811  (T, T, A, T, G, A, G, C, T, C, C, T, T, A, T, ...   \n",
       "812  (T, T, T, A, T, G, A, G, C, T, C, C, T, T, A, ...   \n",
       "816  (T, C, T, T, T, T, T, A, T, G, A, G, C, T, C, ...   \n",
       "\n",
       "                                                Probe1  \\\n",
       "20   (C, C, T, C, C, T, C, C, A, T, G, C, C, G, G, ...   \n",
       "21   (A, C, C, T, C, C, T, C, C, A, T, G, C, C, G, ...   \n",
       "22   (G, A, C, C, T, C, C, T, C, C, A, T, G, C, C, ...   \n",
       "23   (G, G, A, C, C, T, C, C, T, C, C, A, T, G, C, ...   \n",
       "24   (G, G, G, A, C, C, T, C, C, T, C, C, A, T, G, ...   \n",
       "..                                                 ...   \n",
       "809  (C, T, G, G, C, A, T, G, C, T, G, T, C, T, T, ...   \n",
       "810  (T, C, T, G, G, C, A, T, G, C, T, G, T, C, T, ...   \n",
       "811  (T, T, C, T, G, G, C, A, T, G, C, T, G, T, C, ...   \n",
       "812  (T, T, T, C, T, G, G, C, A, T, G, C, T, G, T, ...   \n",
       "816  (T, C, A, G, T, T, T, C, T, G, G, C, A, T, G, ...   \n",
       "\n",
       "                                                Probe2 basicFilter1  \\\n",
       "20   (T, G, C, A, G, T, T, C, T, G, A, A, C, C, C, ...         PASS   \n",
       "21   (C, T, G, C, A, G, T, T, C, T, G, A, A, C, C, ...         PASS   \n",
       "22   (A, C, T, G, C, A, G, T, T, C, T, G, A, A, C, ...         PASS   \n",
       "23   (C, A, C, T, G, C, A, G, T, T, C, T, G, A, A, ...         PASS   \n",
       "24   (C, C, A, C, T, G, C, A, G, T, T, C, T, G, A, ...         PASS   \n",
       "..                                                 ...          ...   \n",
       "809  (A, T, G, A, G, C, T, C, C, T, T, A, T, C, T, ...         PASS   \n",
       "810  (T, A, T, G, A, G, C, T, C, C, T, T, A, T, C, ...         PASS   \n",
       "811  (T, T, A, T, G, A, G, C, T, C, C, T, T, A, T, ...         PASS   \n",
       "812  (T, T, T, A, T, G, A, G, C, T, C, C, T, T, A, ...         PASS   \n",
       "816  (T, C, T, T, T, T, T, A, T, G, A, G, C, T, C, ...         PASS   \n",
       "\n",
       "    basicFilter2  \n",
       "20          PASS  \n",
       "21          PASS  \n",
       "22          PASS  \n",
       "23          PASS  \n",
       "24          PASS  \n",
       "..           ...  \n",
       "809         PASS  \n",
       "810         PASS  \n",
       "811         PASS  \n",
       "812         PASS  \n",
       "816         PASS  \n",
       "\n",
       "[551 rows x 5 columns]"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Probes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa094f59-20b9-420c-9e78-3a15c8f4810d",
   "metadata": {},
   "source": [
    "### 3.和Ref_seq进行比对"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "37dd7b10-32f4-4dbe-aa98-df5cdde6f280",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def write_to_fasta(Probes, fastafile):\n",
    "    \n",
    "    with open(fastafile, \"w\") as handle:\n",
    "        for srec in Probes:\n",
    "            handle.write(f\">{srec.id}\\n\")\n",
    "            handle.write(f\"{str(srec.seq)}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "541cf9d0-38d5-4707-923b-49d428c78576",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fastafile = \"./TMP/region_candidates.fasta\"\n",
    "alignfile = \"./TMP/region_candidates_alignment.tsv\"\n",
    "\n",
    "write_to_fasta(Probes[\"Region\"].to_list(), fastafile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "f51fd078-598a-4e85-8958-f5db1a0d28c4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import subprocess\n",
    "    \n",
    "def blastn(fastafile, result_path, evalue = 5, taxids = \"10090\", strand = 'minus'):\n",
    "    \"\"\"\n",
    "    runs blastn to create alignment results from a fasta file\n",
    "    \"\"\"\n",
    "    call = [\n",
    "        '~/software/ncbi-blast-2.15.0+/bin/blastn',\n",
    "        \"-query\", fastafile,\n",
    "        \"-db\", '../blastndb/refseq_rna',\n",
    "        \"-task\", \"blastn-short\",\n",
    "        \"-evalue\", str(evalue),\n",
    "        \"-strand\", strand,\n",
    "        \"-out\", result_path,\n",
    "        \"-outfmt\", r'\"6 qseqid sacc pident length mismatch qcovs evalue qseq sseq stitle\"',\n",
    "        \"-taxids\", taxids,\n",
    "        \"-num_threads\", '16'\n",
    "        ]\n",
    "    \n",
    "    subprocess.run(\" \".join(call), shell = True, check = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "f3dcaf1f-35e1-442e-9e50-392a8ddb4590",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "blastn(fastafile = fastafile, result_path = alignfile, taxids = '10090', evalue = 100, strand = 'minus')\n",
    "alignedProbe = pd.read_csv(alignfile, sep = \"\\t\",\n",
    "                       names=[\"qseqid\",\"sacc\",\"pident\",\"length\",\"mismatch\",\"qcovs\",\"evalue\", \"qseq\", \"sseq\",\"stitle\"],\n",
    "                           index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "03eeb92c-1c03-4884-a109-f05046e3db13",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def non_specific_check(probe, alignedProbe):\n",
    "        \n",
    "    pid = probe.id\n",
    "    \n",
    "    try:\n",
    "        sub_df = alignedProbe.loc[pid,:].copy()\n",
    "    except:\n",
    "        return 100\n",
    "        \n",
    "    if isinstance(sub_df,pd.core.series.Series):\n",
    "        \n",
    "        return sub_df['evalue']\n",
    "                \n",
    "    else:\n",
    "            \n",
    "        return sub_df['evalue'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "12e2eaaf-cb90-42a6-9fc9-c7d84c8fe138",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "Probes[\"BLASTn_score\"] = Probes[\"Region\"].parallel_apply(lambda x:non_specific_check(x, alignedProbe = alignedProbe))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "4a07f9e0-ddfd-43a4-a51f-574b4ea8e232",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>Probe1</th>\n",
       "      <th>Probe2</th>\n",
       "      <th>basicFilter1</th>\n",
       "      <th>basicFilter2</th>\n",
       "      <th>BLASTn_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>(G, T, T, T, C, T, G, G, C, A, T, G, C, T, G, ...</td>\n",
       "      <td>(C, G, T, T, T, C, T, T, G, T, C, C, C, T, C, ...</td>\n",
       "      <td>(G, T, T, T, C, T, G, G, C, A, T, G, C, T, G, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>48.541358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>779</th>\n",
       "      <td>(T, T, C, T, G, G, C, A, T, G, C, T, G, T, C, ...</td>\n",
       "      <td>(T, T, T, C, T, T, G, T, C, C, C, T, C, A, A, ...</td>\n",
       "      <td>(T, T, C, T, G, G, C, A, T, G, C, T, G, T, C, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>48.619527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780</th>\n",
       "      <td>(T, T, T, C, T, G, G, C, A, T, G, C, T, G, T, ...</td>\n",
       "      <td>(G, T, T, T, C, T, T, G, T, C, C, C, T, C, A, ...</td>\n",
       "      <td>(T, T, T, C, T, G, G, C, A, T, G, C, T, G, T, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>48.697546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777</th>\n",
       "      <td>(C, T, G, G, C, A, T, G, C, T, G, T, C, T, T, ...</td>\n",
       "      <td>(T, C, T, T, G, T, C, C, C, T, C, A, A, C, C, ...</td>\n",
       "      <td>(C, T, G, G, C, A, T, G, C, T, G, T, C, T, T, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>48.851250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>782</th>\n",
       "      <td>(A, G, T, T, T, C, T, G, G, C, A, T, G, C, T, ...</td>\n",
       "      <td>(A, C, G, T, T, T, C, T, T, G, T, C, C, C, T, ...</td>\n",
       "      <td>(A, G, T, T, T, C, T, G, G, C, A, T, G, C, T, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>48.851829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>(A, G, C, C, T, G, C, T, C, C, C, G, G, A, G, ...</td>\n",
       "      <td>(T, G, C, C, A, T, G, C, T, C, C, A, C, T, C, ...</td>\n",
       "      <td>(A, G, C, C, T, G, C, T, C, C, C, G, G, A, G, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>72.724370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>(G, C, C, T, G, C, T, C, C, C, G, G, A, G, C, ...</td>\n",
       "      <td>(G, C, C, A, T, G, C, T, C, C, A, C, T, C, T, ...</td>\n",
       "      <td>(G, C, C, T, G, C, T, C, C, C, G, G, A, G, C, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>72.724370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>(A, A, G, C, C, T, G, C, T, C, C, C, G, G, A, ...</td>\n",
       "      <td>(C, T, G, C, C, A, T, G, C, T, C, C, A, C, T, ...</td>\n",
       "      <td>(A, A, G, C, C, T, G, C, T, C, C, C, G, G, A, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>72.729707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>(G, A, A, G, C, C, T, G, C, T, C, C, C, G, G, ...</td>\n",
       "      <td>(G, C, T, G, C, C, A, T, G, C, T, C, C, A, C, ...</td>\n",
       "      <td>(G, A, A, G, C, C, T, G, C, T, C, C, C, G, G, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>72.740249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>(A, G, A, A, G, C, C, T, G, C, T, C, C, C, G, ...</td>\n",
       "      <td>(A, G, C, T, G, C, C, A, T, G, C, T, C, C, A, ...</td>\n",
       "      <td>(A, G, A, A, G, C, C, T, G, C, T, C, C, C, G, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>72.780723</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>551 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Region  \\\n",
       "781  (G, T, T, T, C, T, G, G, C, A, T, G, C, T, G, ...   \n",
       "779  (T, T, C, T, G, G, C, A, T, G, C, T, G, T, C, ...   \n",
       "780  (T, T, T, C, T, G, G, C, A, T, G, C, T, G, T, ...   \n",
       "777  (C, T, G, G, C, A, T, G, C, T, G, T, C, T, T, ...   \n",
       "782  (A, G, T, T, T, C, T, G, G, C, A, T, G, C, T, ...   \n",
       "..                                                 ...   \n",
       "314  (A, G, C, C, T, G, C, T, C, C, C, G, G, A, G, ...   \n",
       "313  (G, C, C, T, G, C, T, C, C, C, G, G, A, G, C, ...   \n",
       "315  (A, A, G, C, C, T, G, C, T, C, C, C, G, G, A, ...   \n",
       "316  (G, A, A, G, C, C, T, G, C, T, C, C, C, G, G, ...   \n",
       "317  (A, G, A, A, G, C, C, T, G, C, T, C, C, C, G, ...   \n",
       "\n",
       "                                                Probe1  \\\n",
       "781  (C, G, T, T, T, C, T, T, G, T, C, C, C, T, C, ...   \n",
       "779  (T, T, T, C, T, T, G, T, C, C, C, T, C, A, A, ...   \n",
       "780  (G, T, T, T, C, T, T, G, T, C, C, C, T, C, A, ...   \n",
       "777  (T, C, T, T, G, T, C, C, C, T, C, A, A, C, C, ...   \n",
       "782  (A, C, G, T, T, T, C, T, T, G, T, C, C, C, T, ...   \n",
       "..                                                 ...   \n",
       "314  (T, G, C, C, A, T, G, C, T, C, C, A, C, T, C, ...   \n",
       "313  (G, C, C, A, T, G, C, T, C, C, A, C, T, C, T, ...   \n",
       "315  (C, T, G, C, C, A, T, G, C, T, C, C, A, C, T, ...   \n",
       "316  (G, C, T, G, C, C, A, T, G, C, T, C, C, A, C, ...   \n",
       "317  (A, G, C, T, G, C, C, A, T, G, C, T, C, C, A, ...   \n",
       "\n",
       "                                                Probe2 basicFilter1  \\\n",
       "781  (G, T, T, T, C, T, G, G, C, A, T, G, C, T, G, ...         PASS   \n",
       "779  (T, T, C, T, G, G, C, A, T, G, C, T, G, T, C, ...         PASS   \n",
       "780  (T, T, T, C, T, G, G, C, A, T, G, C, T, G, T, ...         PASS   \n",
       "777  (C, T, G, G, C, A, T, G, C, T, G, T, C, T, T, ...         PASS   \n",
       "782  (A, G, T, T, T, C, T, G, G, C, A, T, G, C, T, ...         PASS   \n",
       "..                                                 ...          ...   \n",
       "314  (A, G, C, C, T, G, C, T, C, C, C, G, G, A, G, ...         PASS   \n",
       "313  (G, C, C, T, G, C, T, C, C, C, G, G, A, G, C, ...         PASS   \n",
       "315  (A, A, G, C, C, T, G, C, T, C, C, C, G, G, A, ...         PASS   \n",
       "316  (G, A, A, G, C, C, T, G, C, T, C, C, C, G, G, ...         PASS   \n",
       "317  (A, G, A, A, G, C, C, T, G, C, T, C, C, C, G, ...         PASS   \n",
       "\n",
       "    basicFilter2  BLASTn_score  \n",
       "781         PASS     48.541358  \n",
       "779         PASS     48.619527  \n",
       "780         PASS     48.697546  \n",
       "777         PASS     48.851250  \n",
       "782         PASS     48.851829  \n",
       "..           ...           ...  \n",
       "314         PASS     72.724370  \n",
       "313         PASS     72.724370  \n",
       "315         PASS     72.729707  \n",
       "316         PASS     72.740249  \n",
       "317         PASS     72.780723  \n",
       "\n",
       "[551 rows x 6 columns]"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Probes.sort_values('BLASTn_score')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f19cdc8-c290-4dd3-adc8-14476deee700",
   "metadata": {},
   "source": [
    "### 4.和rmsk进行比对"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "db739502-8444-4387-bbd3-b9f55bf3dcce",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "bowtie2file = \"./TMP/region_candidates_alignment.sam\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "d81ecf91-a320-4fba-ab6e-d4e20ab679b5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def bowtie2(fastafile:str, result_path:str):\n",
    " \n",
    "    db = '/media/duan/sda2/Reference_DataBase/mm39/bowtie2index/GRCm39'\n",
    "           \n",
    "    call = [\n",
    "        'bowtie2',\n",
    "        '-x', db,\n",
    "        '-U', fastafile,\n",
    "        '-S', result_path,\n",
    "        '-f',\n",
    "        '--local',\n",
    "        '-D 20 -R 3 -N 1 -L 20 -i S,1,0.50',\n",
    "        '-a',\n",
    "        '--reorder',\n",
    "        '-p', '16'\n",
    "    ]\n",
    "    \n",
    "    cp = subprocess.run(\" \".join(call), shell = True, check = True, capture_output=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "b1405712-3ce3-48e6-b1ae-20ede162186d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "bowtie2(fastafile = fastafile, result_path = bowtie2file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "2c02add8-6735-4866-a5f4-20c1c7a10ff7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def bedtools_intersect_filter(bowtie2file, TE_INPUT, key):\n",
    "    \n",
    "    import subprocess\n",
    "    \n",
    "    # 先将sam文件压缩成bam\n",
    "    \n",
    "    command = f'samtools view -Sb {bowtie2file} > ./TMP/region_candidates_alignment.bam'\n",
    "    cp = subprocess.run(command, shell = True, check = True, capture_output = True)\n",
    "    \n",
    "    # 然后进行intersect\n",
    "    \n",
    "    rmsk_bed = '/media/duan/sda2/Reference_DataBase/mm39/GRCm39_GENCODE_rmsk_TE.bed'\n",
    "    command = f\"bedtools intersect -a ./TMP/region_candidates_alignment.bam -b {rmsk_bed} -wa -wb -bed > ./TMP/region_candidates_alignment.bed\"\n",
    "    cp = subprocess.run(command, shell = True, check = True, capture_output = True)\n",
    "    \n",
    "    Bowtie2_On_score = Counter()\n",
    "    Bowtie2_Off_score = Counter()\n",
    "    \n",
    "    with open(\"./TMP/region_candidates_alignment.bed\", 'r') as handle:\n",
    "        for line in handle.readlines():\n",
    "            parsed_line = line.strip().split(\"\\t\")\n",
    "            subject = \"|\".join(parsed_line[0:3]) + f\"|{parsed_line[5]}\"\n",
    "            if subject in TE_INPUT[key]:\n",
    "                Bowtie2_On_score[parsed_line[3]] += 1\n",
    "            else:\n",
    "                Bowtie2_Off_score[parsed_line[3]] += 1\n",
    "                \n",
    "    return Bowtie2_On_score, Bowtie2_Off_score"
   ]
  },
  {
   "cell_type": "raw",
   "id": "1d822f08-499c-4c12-9f23-a22a3f15e633",
   "metadata": {
    "tags": []
   },
   "source": [
    "bedtools_intersect_filter(bowtie2file, TE_INPUT, key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "04757cd2-1e83-480e-9c18-de45f66f00ad",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "Bowtie2_On_score, Bowtie2_Off_score = bedtools_intersect_filter(bowtie2file, TE_INPUT, key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "d6708133-3757-4d5b-b5e9-43f33af05ed9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "Off_list = []\n",
    "On_list = []\n",
    "for x in Probes['Region']:\n",
    "    try:\n",
    "        tmp = Bowtie2_Off_score[x.id]\n",
    "    except KeyError:\n",
    "        Off_list.append(0)\n",
    "        \n",
    "    Off_list.append(tmp)\n",
    "    \n",
    "    try:\n",
    "        tmp = Bowtie2_On_score[x.id]\n",
    "    except KeyError:\n",
    "        On_list.append(0)\n",
    "    \n",
    "    On_list.append(tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "d29fe0d8-7439-4a20-9c35-c9d36a9f0bd8",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "Probes['Bowtie2_On_Score'] = On_list\n",
    "Probes['Bowtie2_Off_Score'] = Off_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "35b17728-119b-45bd-8fbf-7863c8bbbe2b",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>Probe1</th>\n",
       "      <th>Probe2</th>\n",
       "      <th>basicFilter1</th>\n",
       "      <th>basicFilter2</th>\n",
       "      <th>BLASTn_score</th>\n",
       "      <th>Bowtie2_On_Score</th>\n",
       "      <th>Bowtie2_Off_Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>(T, G, C, A, G, T, T, C, T, G, A, A, C, C, C, ...</td>\n",
       "      <td>(C, C, T, C, C, T, C, C, A, T, G, C, C, G, G, ...</td>\n",
       "      <td>(T, G, C, A, G, T, T, C, T, G, A, A, C, C, C, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>65.288772</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>(C, T, G, C, A, G, T, T, C, T, G, A, A, C, C, ...</td>\n",
       "      <td>(A, C, C, T, C, C, T, C, C, A, T, G, C, C, G, ...</td>\n",
       "      <td>(C, T, G, C, A, G, T, T, C, T, G, A, A, C, C, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>65.494302</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>(A, C, T, G, C, A, G, T, T, C, T, G, A, A, C, ...</td>\n",
       "      <td>(G, A, C, C, T, C, C, T, C, C, A, T, G, C, C, ...</td>\n",
       "      <td>(A, C, T, G, C, A, G, T, T, C, T, G, A, A, C, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>66.002075</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>(C, A, C, T, G, C, A, G, T, T, C, T, G, A, A, ...</td>\n",
       "      <td>(G, G, A, C, C, T, C, C, T, C, C, A, T, G, C, ...</td>\n",
       "      <td>(C, A, C, T, G, C, A, G, T, T, C, T, G, A, A, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>65.181174</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>(C, C, A, C, T, G, C, A, G, T, T, C, T, G, A, ...</td>\n",
       "      <td>(G, G, G, A, C, C, T, C, C, T, C, C, A, T, G, ...</td>\n",
       "      <td>(C, C, A, C, T, G, C, A, G, T, T, C, T, G, A, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>65.793591</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809</th>\n",
       "      <td>(A, T, G, A, G, C, T, C, C, T, T, A, T, C, T, ...</td>\n",
       "      <td>(C, T, G, G, C, A, T, G, C, T, G, T, C, T, T, ...</td>\n",
       "      <td>(A, T, G, A, G, C, T, C, C, T, T, A, T, C, T, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>59.045402</td>\n",
       "      <td>0</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>810</th>\n",
       "      <td>(T, A, T, G, A, G, C, T, C, C, T, T, A, T, C, ...</td>\n",
       "      <td>(T, C, T, G, G, C, A, T, G, C, T, G, T, C, T, ...</td>\n",
       "      <td>(T, A, T, G, A, G, C, T, C, C, T, T, A, T, C, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>59.110857</td>\n",
       "      <td>0</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811</th>\n",
       "      <td>(T, T, A, T, G, A, G, C, T, C, C, T, T, A, T, ...</td>\n",
       "      <td>(T, T, C, T, G, G, C, A, T, G, C, T, G, T, C, ...</td>\n",
       "      <td>(T, T, A, T, G, A, G, C, T, C, C, T, T, A, T, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>59.739429</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>(T, T, T, A, T, G, A, G, C, T, C, C, T, T, A, ...</td>\n",
       "      <td>(T, T, T, C, T, G, G, C, A, T, G, C, T, G, T, ...</td>\n",
       "      <td>(T, T, T, A, T, G, A, G, C, T, C, C, T, T, A, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>60.211299</td>\n",
       "      <td>0</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816</th>\n",
       "      <td>(T, C, T, T, T, T, T, A, T, G, A, G, C, T, C, ...</td>\n",
       "      <td>(T, C, A, G, T, T, T, C, T, G, G, C, A, T, G, ...</td>\n",
       "      <td>(T, C, T, T, T, T, T, A, T, G, A, G, C, T, C, ...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>59.402367</td>\n",
       "      <td>0</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>551 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Region  \\\n",
       "20   (T, G, C, A, G, T, T, C, T, G, A, A, C, C, C, ...   \n",
       "21   (C, T, G, C, A, G, T, T, C, T, G, A, A, C, C, ...   \n",
       "22   (A, C, T, G, C, A, G, T, T, C, T, G, A, A, C, ...   \n",
       "23   (C, A, C, T, G, C, A, G, T, T, C, T, G, A, A, ...   \n",
       "24   (C, C, A, C, T, G, C, A, G, T, T, C, T, G, A, ...   \n",
       "..                                                 ...   \n",
       "809  (A, T, G, A, G, C, T, C, C, T, T, A, T, C, T, ...   \n",
       "810  (T, A, T, G, A, G, C, T, C, C, T, T, A, T, C, ...   \n",
       "811  (T, T, A, T, G, A, G, C, T, C, C, T, T, A, T, ...   \n",
       "812  (T, T, T, A, T, G, A, G, C, T, C, C, T, T, A, ...   \n",
       "816  (T, C, T, T, T, T, T, A, T, G, A, G, C, T, C, ...   \n",
       "\n",
       "                                                Probe1  \\\n",
       "20   (C, C, T, C, C, T, C, C, A, T, G, C, C, G, G, ...   \n",
       "21   (A, C, C, T, C, C, T, C, C, A, T, G, C, C, G, ...   \n",
       "22   (G, A, C, C, T, C, C, T, C, C, A, T, G, C, C, ...   \n",
       "23   (G, G, A, C, C, T, C, C, T, C, C, A, T, G, C, ...   \n",
       "24   (G, G, G, A, C, C, T, C, C, T, C, C, A, T, G, ...   \n",
       "..                                                 ...   \n",
       "809  (C, T, G, G, C, A, T, G, C, T, G, T, C, T, T, ...   \n",
       "810  (T, C, T, G, G, C, A, T, G, C, T, G, T, C, T, ...   \n",
       "811  (T, T, C, T, G, G, C, A, T, G, C, T, G, T, C, ...   \n",
       "812  (T, T, T, C, T, G, G, C, A, T, G, C, T, G, T, ...   \n",
       "816  (T, C, A, G, T, T, T, C, T, G, G, C, A, T, G, ...   \n",
       "\n",
       "                                                Probe2 basicFilter1  \\\n",
       "20   (T, G, C, A, G, T, T, C, T, G, A, A, C, C, C, ...         PASS   \n",
       "21   (C, T, G, C, A, G, T, T, C, T, G, A, A, C, C, ...         PASS   \n",
       "22   (A, C, T, G, C, A, G, T, T, C, T, G, A, A, C, ...         PASS   \n",
       "23   (C, A, C, T, G, C, A, G, T, T, C, T, G, A, A, ...         PASS   \n",
       "24   (C, C, A, C, T, G, C, A, G, T, T, C, T, G, A, ...         PASS   \n",
       "..                                                 ...          ...   \n",
       "809  (A, T, G, A, G, C, T, C, C, T, T, A, T, C, T, ...         PASS   \n",
       "810  (T, A, T, G, A, G, C, T, C, C, T, T, A, T, C, ...         PASS   \n",
       "811  (T, T, A, T, G, A, G, C, T, C, C, T, T, A, T, ...         PASS   \n",
       "812  (T, T, T, A, T, G, A, G, C, T, C, C, T, T, A, ...         PASS   \n",
       "816  (T, C, T, T, T, T, T, A, T, G, A, G, C, T, C, ...         PASS   \n",
       "\n",
       "    basicFilter2  BLASTn_score  Bowtie2_On_Score  Bowtie2_Off_Score  \n",
       "20          PASS     65.288772                 0                 27  \n",
       "21          PASS     65.494302                 0                 29  \n",
       "22          PASS     66.002075                 0                 29  \n",
       "23          PASS     65.181174                 0                 29  \n",
       "24          PASS     65.793591                 0                 29  \n",
       "..           ...           ...               ...                ...  \n",
       "809         PASS     59.045402                 0                 52  \n",
       "810         PASS     59.110857                 0                 59  \n",
       "811         PASS     59.739429                 0                 60  \n",
       "812         PASS     60.211299                 0                 58  \n",
       "816         PASS     59.402367                 0                 59  \n",
       "\n",
       "[551 rows x 8 columns]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Probes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36ce296c-ff77-4a8f-9eac-15e5252a87ba",
   "metadata": {},
   "source": [
    "### 5.拼接成完整序列"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "06c1ba17-40f9-46d1-9746-8669810e3a16",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_bridge_seq(bridge_id):\n",
    "        \n",
    "    bseqs = pd.read_csv('./BridgeSequence/Bridges_used_for_assemble_20240704.tsv', sep=\"\\t\",index_col=0)\n",
    "    return bseqs.loc[bridge_id,\"Seq\"]\n",
    "\n",
    "bridge_seq = get_bridge_seq(bridge_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "9c84458e-a7e2-4b42-bcf8-3c70ab1101b6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_full_probes1(p1, bridge_seq):\n",
    "    \n",
    "    \"\"\"\n",
    "    该函数用于拼接完整的p1序列\n",
    "    \"\"\"\n",
    "    return bridge_seq[0:14] + \"ta\" + p1\n",
    "\n",
    "def get_full_probes2(p2, bridge_seq):\n",
    "    \n",
    "    \"\"\"\n",
    "    该函数用于拼接完整的p2序列\n",
    "    \"\"\"\n",
    "    return p2 + \"ta\" + bridge_seq[-14:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "4cfa2ff9-8acc-4207-91d6-6ebf9d9d9d68",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "Probes[\"Full_Probe1\"] = Probes[\"Probe1\"].parallel_apply(lambda x:get_full_probes1(x, bridge_seq = bridge_seq))\n",
    "Probes[\"Full_Probe2\"] = Probes[\"Probe2\"].parallel_apply(lambda x:get_full_probes2(x, bridge_seq = bridge_seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ade1a2ec-5a32-4816-aade-e760148ea548",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 6.进行打分"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "d31dce59-c128-4514-a92f-e49ce8355337",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Probes = Probes[Probes['Bowtie2_On_Score'] > 0].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "473e76c5-5577-426c-bae2-419808d0bdb9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "2ab4f2f3-fe99-4a76-8827-e8388e731dd9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_evalue(row):\n",
    "    \n",
    "\n",
    "    part1 = 4 / (row['BLASTn_score'] + 1)\n",
    "    \n",
    "    part2 = 8 / max([row['Bowtie2_On_Score'], 1]) # 0 - 8\n",
    "\n",
    "    part3 = min([np.log10(row['Bowtie2_Off_Score'] + 1), 4]) # 0 - 4\n",
    "    \n",
    "    return part1 + part2 + part3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "bbfdc7d0-952e-4b1a-be53-6e611bc2f08e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "Probes['Evaluation'] = Probes.parallel_apply(lambda row: get_evalue(row), axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "972115b1-50a2-451e-b9c0-b1361f19c4fc",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "Probes = Probes.sort_values('Evaluation')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "c0d13b3d-eb56-4008-a795-936241fb26c6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "result = pd.DataFrame({\n",
    "        'name': key,\n",
    "        'probe_name': [x.id for x in Probes[\"Region\"]],\n",
    "        'position':[x.description for x in Probes[\"Region\"]],\n",
    "        'probe A':[str(x.seq) for x in Probes[\"Full_Probe1\"]],\n",
    "        'probe B':[str(x.seq) for x in Probes[\"Full_Probe2\"]],\n",
    "        \"Evaluation\": Probes['Evaluation'],\n",
    "        \"BLASTn_score\": Probes['BLASTn_score'],\n",
    "        \"Bowtie2_On_Score\": Probes['Bowtie2_On_Score'],\n",
    "        \"Bowtie2_Off_Score\": Probes['Bowtie2_Off_Score'],\n",
    "        'bridge_id': bridge_id,\n",
    "        'probe_strand': [x.id.split(\"_\")[-1] for x in Probes[\"Region\"]]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "e7557f0b-8b6c-4352-b1be-1551732a7292",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "result = result.reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "2a9aff8b-1bad-4671-9a10-05601f51d3b0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "result.to_csv(f\"{key}_result.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd6cecdf-1309-4d66-b45b-a8d766ad108b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "b46d0747-2849-4309-bf62-a9727c9bf014",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "KEEP_INDICES = [\n",
    "    10,46,49,137,\n",
    "] # input index of selected probes manually\n",
    "result = result.loc[KEEP_INDICES, :].copy()\n",
    "result = result.reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "27b79a03-76c9-486f-b773-642e1094886f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def generate_sangong(data):\n",
    "    # This Block is used to generate name + sequence sangong accepted.\n",
    "    # The name format is \"[GENE]_[ProbePosition]_[ProbeNumber][ProbeLength]_[PadNumber]\", such as H2-Ab1_506_1A30_7196.\n",
    "    final_name = []\n",
    "    final_primer = []\n",
    "    gene_check = []\n",
    "    Final_gene = []\n",
    "    for i in range(len(data)):\n",
    "        pos1 = data.loc[i,\"name\"]\n",
    "        gene_check.append(pos1)\n",
    "        number_ = [j for j in gene_check if j == pos1]\n",
    "        number = len(number_)\n",
    "        \n",
    "        bridge_id = data.loc[i,\"bridge_id\"]\n",
    "        if bridge_id.startswith(\"Pad\"):\n",
    "            pos5 = \"P\" + bridge_id.split(\"_\")[-1]\n",
    "        else:\n",
    "            try:\n",
    "                pos5 = bridge_id.split(\"seq\")[1]\n",
    "            except:\n",
    "                pos5 = bridge_id.split(\"bridge\")[1]\n",
    "            \n",
    "        pos2 = str(int(data.loc[i,\"position\"]))\n",
    "        pos3 = data.loc[i,\"probe A\"]\n",
    "        pos4 = data.loc[i,\"probe B\"]\n",
    "        final1 = pos1+\"_\"+pos2+\"_\" + str(number) +\"A\" + \"_\" + pos5\n",
    "        final2 = pos1+\"_\"+pos2+\"_\" + str(number) +\"B\" + \"_\" + pos5\n",
    "\n",
    "        final_name.append(final1)\n",
    "        final_name.append(final2)\n",
    "        final_primer.append(pos3)\n",
    "        final_primer.append(pos4)\n",
    "        Final_gene.append(pos1)\n",
    "        Final_gene.append(pos1)\n",
    "\n",
    "    print(\"There are {} sequences in total.\".format(str(len(final_name))))\n",
    "    final_df = pd.DataFrame({\"name\":final_name, \"seq\":final_primer,\"gene\":Final_gene})\n",
    "    \n",
    "    return final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "5bbce59e-29ef-48d1-bc31-7254f066d9e6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 8 sequences in total.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>seq</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IAPLTR3-int_358_1A_P30330</td>\n",
       "      <td>CAGTAACGTTCGCAtaTCCCAAAACAGAAGCCTGCTCCCGGAGCAG</td>\n",
       "      <td>IAPLTR3-int</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IAPLTR3-int_358_1B_P30330</td>\n",
       "      <td>AGAAAAAGTCTGTTCTCAAAACAAAAGCCTtaATCAGGGTTACAAT</td>\n",
       "      <td>IAPLTR3-int</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IAPLTR3-int_251_2A_P30330</td>\n",
       "      <td>CAGTAACGTTCGCAtaACGTAGGACTGACCGCACATCAAGGTCACA</td>\n",
       "      <td>IAPLTR3-int</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IAPLTR3-int_251_2B_P30330</td>\n",
       "      <td>CTGACTGAGCTTGTTAGTGCTGCCGTTTATtaATCAGGGTTACAAT</td>\n",
       "      <td>IAPLTR3-int</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IAPLTR3-int_67_3A_P30330</td>\n",
       "      <td>CAGTAACGTTCGCAtaACTTACGATTTCCACTGCAGTTCTGAACCC</td>\n",
       "      <td>IAPLTR3-int</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>IAPLTR3-int_67_3B_P30330</td>\n",
       "      <td>CAACAGGCTCACAGTTAAGCATGTCTCTCTtaATCAGGGTTACAAT</td>\n",
       "      <td>IAPLTR3-int</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>IAPLTR3-int_635_4A_P30330</td>\n",
       "      <td>CAGTAACGTTCGCAtaCTGTGGTCAAGGAACCTGAGCAAGCATACC</td>\n",
       "      <td>IAPLTR3-int</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>IAPLTR3-int_635_4B_P30330</td>\n",
       "      <td>ATCTACGTCTCTTCCTAACTTGTTCCACGAtaATCAGGGTTACAAT</td>\n",
       "      <td>IAPLTR3-int</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        name                                             seq  \\\n",
       "0  IAPLTR3-int_358_1A_P30330  CAGTAACGTTCGCAtaTCCCAAAACAGAAGCCTGCTCCCGGAGCAG   \n",
       "1  IAPLTR3-int_358_1B_P30330  AGAAAAAGTCTGTTCTCAAAACAAAAGCCTtaATCAGGGTTACAAT   \n",
       "2  IAPLTR3-int_251_2A_P30330  CAGTAACGTTCGCAtaACGTAGGACTGACCGCACATCAAGGTCACA   \n",
       "3  IAPLTR3-int_251_2B_P30330  CTGACTGAGCTTGTTAGTGCTGCCGTTTATtaATCAGGGTTACAAT   \n",
       "4   IAPLTR3-int_67_3A_P30330  CAGTAACGTTCGCAtaACTTACGATTTCCACTGCAGTTCTGAACCC   \n",
       "5   IAPLTR3-int_67_3B_P30330  CAACAGGCTCACAGTTAAGCATGTCTCTCTtaATCAGGGTTACAAT   \n",
       "6  IAPLTR3-int_635_4A_P30330  CAGTAACGTTCGCAtaCTGTGGTCAAGGAACCTGAGCAAGCATACC   \n",
       "7  IAPLTR3-int_635_4B_P30330  ATCTACGTCTCTTCCTAACTTGTTCCACGAtaATCAGGGTTACAAT   \n",
       "\n",
       "          gene  \n",
       "0  IAPLTR3-int  \n",
       "1  IAPLTR3-int  \n",
       "2  IAPLTR3-int  \n",
       "3  IAPLTR3-int  \n",
       "4  IAPLTR3-int  \n",
       "5  IAPLTR3-int  \n",
       "6  IAPLTR3-int  \n",
       "7  IAPLTR3-int  "
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df = generate_sangong(result)\n",
    "final_df.to_csv(f\"{key}_sangon.csv\", sep=\",\", index = False)\n",
    "final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "218d01dd-bfc7-4927-af8a-eb5097d774d1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:STalign]",
   "language": "python",
   "name": "conda-env-STalign-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
